provider/adapters/src/fallback/adapter.rs

```diff
@@ -3,25 +3,186 @@
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
 use super::*;
+use crate::helpers::result_is_err_missing_resource_options;
 
-pub trait LocaleFallbackProvider {
-    fn load_fallback_metadata_for_key(
-        &self,
-        key: ResourceKey,
-    ) -> Result<LocaleFallbackKeyMetadata, DataError>;
+/// A data provider wrapper that performs locale fallback. This enables arbitrary locales to be
+/// handled at runtime.
+///
+/// # Examples
+///
+/// ```
+/// use icu_locid::locale;
+/// use icu_provider::prelude::*;
+/// use icu_plurals::PluralRules;
+/// use icu_provider_adapters::fallback::LocaleFallbackProvider;
+///
+/// let provider = icu_testdata::get_provider();
+///
+/// // The provider does not have data for "ar-SA":
+/// PluralRules::try_new_cardinal(locale!("ar-SA"), &provider)
+///     .map(|_| ())
+///     .expect_err("No data for ar-SA");
+///
+/// // But if we wrap the provider in a fallback provider...
+/// let provider = LocaleFallbackProvider::try_new(provider)
+///     .expect("Fallback data present");
+///
+/// // ...then we can load "ar-SA" based on "ar" data
+/// let pr = PluralRules::try_new_cardinal(locale!("ar-SA"), &provider)
+///     .expect("Using fallback provider");
+///
+/// assert_eq!(pr.select(5_usize), icu_plurals::PluralCategory::Few);
+/// ```
```

Comment:
Can we use a hello world example? I know it would require using `new_with_fallbacker` because we need to get fallback data from somewhere else, but I think it's nicer to read the Arabic hello world than having to know that 5 is `PluralCategory::Few` in Arabic. It would also avoid the provider -> component dependency.

Reply:
I wrote the example like this because I wanted to use `try_new` instead of `new_with_fallbacker`... but I'll figure something out. I can make an example with `try_new` that loads the buffer without parsing it, and then a `new_with_fallbacker` for a functional example.

Reply:
You could also use `no_run` and pretend the hello world provider contains fallback data.

Reply:
I sent #2173 so I can use the same provider for both fallback data and hello world data. I cherry-picked the commit in here, which should go away if the other PR is merged.